kvredis/src/main.rs

```diff
@@ -0,0 +1,307 @@
+    /// including setting up per-actor resources, and checking authorization.
+    /// If the link is allowed, return true, otherwise return false to deny the link.
+    async fn put_link(&self, ld: &LinkDefinition) -> RpcResult<bool> {
+        let redis_url = match ld.values.get(REDIS_URL_KEY) {
+            Some(v) => v.as_str(),
+            None => DEFAULT_CONNECT_URL,
+        };
+        let client = redis::Client::open(redis_url).map_err(|e| {
+            RpcError::ProviderInit(format!("redis connection to {}: {}", redis_url, e))
+        })?;
+        let connection = client.get_async_connection().await.map_err(|e| {
+            RpcError::ProviderInit(format!("redis connection to {}: {}", redis_url, e))
+        })?;
+
+        let mut update_map = self.actors.write().await;
+        update_map.insert(ld.actor_id.to_string(), RwLock::new(connection));
+
+        Ok(true)
+    }
+
+    /// Handle notification that a link is dropped - close the connection
+    async fn delete_link(&self, actor_id: &str) {
+        let mut aw = self.actors.write().await;
+        if let Some(conn) = aw.remove(actor_id) {
+            log::info!("redis closing connection for actor {}", actor_id);
+            drop(conn)
+        }
+    }
+
+    /// Handle shutdown request by closing all connections
+    async fn shutdown(&self) -> Result<(), Infallible> {
+        let mut aw = self.actors.write().await;
+        // empty the actor link data and stop all servers
+        for (_, conn) in aw.drain() {
+            drop(conn)
+        }
+        Ok(())
+    }
+}
+
+fn to_rpc_err(e: RedisError) -> RpcError {
+    RpcError::Other(format!("redis error: {}", e.to_string()))
+}
+
+// There are two api styles you can use for invoking redis. You can build any raw command
+// as a string command and a sequence of args:
+// ```
+//     let mut cmd = redis::cmd("SREM");
+//     let value: u32 = self.exec(ctx, &mut cmd.arg(&arg.set_name).arg(&arg.value)).await?;
+// ```
+// or you can call a method on Cmd, as in
+// ```
+//     let mut cmd = redis::Cmd::srem(&arg.set_name, &arg.value);
+//     let value: u32 = self.exec(ctx, &mut cmd).await?;
+//```
+// The latter api style has better rust compile-time type checking for args.
+// The rust docs for cmd and Cmd don't document arg types or return types.
+// For that, you need to look at https://redis.io/commands#
+
+/// Handle KeyValue methods that interact with redis
+#[async_trait]
+impl KeyValue for KvRedisProvider {
+    /// Increments a numeric value, returning the new value
+    async fn increment(&self, ctx: &Context, arg: &IncrementRequest) -> RpcResult<i32> {
+        let mut cmd = redis::Cmd::incr(&arg.key, &arg.value);
+        let val: i32 = self.exec(ctx, &mut cmd).await?;
+        Ok(val)
+    }
+
+    /// Returns true if the store contains the key
+    async fn contains<TS: ToString + ?Sized + Sync>(
+        &self,
+        ctx: &Context,
+        arg: &TS,
+    ) -> RpcResult<bool> {
```

Comment:
For this and the other commands, I'm a little confused as to the purpose of the generic TS here. Is there a downside to just using a reference to a string slice `&str` to keep the method signature less complicated?

Reply:
The purpose is to allow you to use &str, String, or &String - it gives the most flexibility for the user of the api. It's all code-generated, so it doesn't add complexity to the developer either.

The secondary benefit, besides convenience for the caller, is reducing an allocation. If we made it &str, and the caller had passed &String, the function has to allocate another String because it needs to be owned further down. On the other hand, if the caller has a string they can pass it and no allocation is required.

Reply:
Okay, the real thing that I think makes this fine is that it's all codegen 👍🏻 Thanks for the answer